ops/testing.py

```diff
@@ -2425,66 +2537,71 @@ def push(
             raise pebble.PathError(
                 'generic-file-error',
                 f'permissions not within 0o000 to 0o777: {permissions:#o}')
+        self.__check_absolute_path(path)
+        file_path = self._root / path[1:]
+        if make_dirs and not file_path.parent.exists():
+            self.make_dir(
+                os.path.split(path)[0],
+                make_parents=True,
+                permissions=None,
+                user_id=user_id,
+                user=user,
+                group_id=group_id,
+                group=group)
+        permissions = permissions if permissions is not None else 0o644
         try:
-            self._fs.create_file(
-                path, source, encoding=encoding, make_dirs=make_dirs, permissions=permissions,
-                user_id=user_id, user=user, group_id=group_id, group=group)
+            if isinstance(source, str):
+                content = source.encode(encoding=encoding)
+            elif isinstance(source, bytes):
+                content = source
+            else:
+                source_read: Union[str, bytes] = source.read()
+                content = (
+                    source_read
+                    if isinstance(source_read, bytes)
+                    else source_read.encode(encoding=encoding)
+                )
+            file_path.write_bytes(content)
+            os.chmod(file_path, permissions)
+            self.__chown(file_path, user_id=user_id, user=user, group_id=group_id, group=group)
         except FileNotFoundError as e:
             raise pebble.PathError(
                 'not-found', f'parent directory not found: {e.args[0]}')
-        except NonAbsolutePathError as e:
-            raise pebble.PathError(
-                'generic-file-error',
-                f'paths must be absolute, got {e.args[0]!r}'
-            )
+        except NotADirectoryError:
+            raise pebble.PathError('generic-file-error',
+                                   f'open {path}.{secrets.token_urlsafe(8)}~: not a directory')
```

Comment:
I don't understand the use of `secrets.token_urlsafe` here. Can you explain? Also, when would the `NotADirectoryError` happen here?

Reply:
When pebble transferring a file, it will first write to a temporary path (filename + random string + ~), and rename to the final destination when the transferring is finished. This is just mimicking the pebble error, while  `secrets.token_urlsafe` is just a quick way to generate a random string.

This error will happen if the parent directory doesn't exist and `make_dirs` is set to `False`.